### ISSUE TITLE
Improved debugger integration

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#554]
_From [remi.fla...@gmail.com](https://code.google.com/u/117263569492313097508/) on 2011-02-10T07:34:39Z_

2 important debugger improvement

-When debugging using pdb, it is not possible to access the current variables in the "Variable explorer". This is a killer matlab feature.

-Could we have pdb buttons shortcut in the editor for the basic tasks 'n' ...

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=554_

**Comment by spyder-issues-migrator** [comment]
_From [asterbom...@gmail.com](https://code.google.com/u/111310574906366460267/) on 2011-03-01T06:22:16Z_

I could not agree more: these are indeed important improvements!!!

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2011-03-21T15:04:56Z_

**Status:** Accepted  
  **Labels:** -Type-Defect Type-Enhancement

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-05-16T01:27:47Z_

**Labels:** Cat-Debugger

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T14:11:31Z_

**Labels:** -Type-Enhancement Type-Enh

**Comment by spyder-issues-migrator** [comment]
_From [DavidAnt...@gmail.com](https://code.google.com/u/108698404730356683634/) on 2012-05-17T19:03:54Z_

I previously had a problem whereby the variable explorer would not show variables in debugging mode. However, I just upgraded from Ubuntu 10.04 to 12.04, and I no longer have this issue. So I guess this issue is most likely a problem with a particular version of Python/Qt/PyQt or some other package.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-05-25T10:29:06Z_

issue #535 has been merged into this issue.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2012-05-25T13:18:40Z_

Yes. Step by step instructions along with version information will make the problem more evident for those like me who never used Spyder with pdb so far.

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2012-05-31T12:55:18Z_

See further discussion on this issue in the discussion group here: https://groups.google.com/d/topic/spyderlib/Yhe6THxNCyY/discussion Also a simple implementation of the debugging toolbar can be found here: https://code.google.com/r/jedludlow-spyderlib-default/source/browse/?name=jedludlow-issue-554

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2012-06-01T12:11:15Z_

Further discussion can be found on this thread, too. https://groups.google.com/d/topic/spyderlib/udt8kfUyxJ0/discussion

**Comment by spyder-issues-migrator** [comment]
_From [steve.f....@gmail.com](https://code.google.com/u/117650558843796505335/) on 2012-06-15T09:05:48Z_

The comment thread redirects back here, so I'll comment here.  I'd like to see a traceback/stack widget added that shows the call stack and can be clicked to move around in the stack.

I'd also like to see the method for starting/running the debugger streamlined.  I had the idea of having an enable checkbox on the breakpoint widget that controls whether pressing F5 starts normal or debug mode.  I was also curious if it would be possible to control flow natively in Spyder instead of using pdb.

I'm finding the integration with pdb quite a bit more reliable than it used to be.  I used to frequently experience Spyder lock ups and weird behavior while attempting to debug.  I have seen this happen a couple of times where the entire editor locks up.  The odd thing is I can click on the "X" to attempt to kill it.  One of the options on Windows 7 is to attempt resume the application, which kind of works, but the socket connection is destroyed and no more input can be sent into the console, so at least it doesn't lock up the editor permanently.  I do have to restart Spyder at this point to actually continue debugging.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2012-06-29T08:54:46Z_

Jed, I tried the toolbar and it works great on our Python interpreter but it doesn't trigger any action on the IPython one. Is it possible to fix this?

Besides, I would like to see the 'Debug' button (the one present in the Run menu) on the toolbar too. I think with these changes is ready to merge. What do you think?

**Owner:** jed.lud...@gmail.com  
  **Labels:** MS-v2.2

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2012-07-02T20:49:02Z_

With regard to IPython, if you are not debugging the toolbar sends the right commands to the IPython console. The problem shows up when pdb has launched then the commands don't get sent it. I have to believe it's because all the special pdb hacks in https://code.google.com/p/spyderlib/source/browse/spyderlib/widgets/externalshell/monitor.py and https://code.google.com/p/spyderlib/source/browse/spyderlib/widgets/externalshell/sitecustomize.py just aren't there in Spyder's IPython console to manage the traffic back and forth.

I guess it is possible to hack in the 'Debug' button to the debugging toolbar, but there are overall structural problems with this. The 'Run' toolbar and all it's actions are really managed by the editor plugin, and the debugging toolbar is all managed by the main spyder module. To get things from the run toolbar onto the debug toolbar requires that you wait for the editor plugin to initialize first and then insert run actions into the debug toolbar. I'm' already doing a little bit of this stuff to make some debugging commands show up in the Run menus, but I don't like it. I'd prefer not to spread the mess any further. Is it not acceptable for a user to simply show the run toolbar and place it next to the debug toolbar if the user wants both icons available in the UI?

**Comment by spyder-issues-migrator** [comment]
_From [steve.f....@gmail.com](https://code.google.com/u/117650558843796505335/) on 2012-12-19T09:13:24Z_

What's the status on getting this into 2.2?

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2012-12-22T10:31:27Z_

I have figured out why the debugging toolbar commands do not work in IPython. When Spyder sends commands to simple Python shells it does so at a very low level, essentially creating characters on stdin. So this mechanism works to send the initial debug request *and* to place pdb commands on the prompt. In contrast, Spyder sends commands to IPython consoles using a higher level method called execute() which is available as part of the standard IPython Qt console app and associated widgets. I asked about how to handle this on the IPtython dev list: http://mail.scipy.org/pipermail/ipython-dev/2012-December/010945.html When IPython enters debug mode it is no longer accepting "execute" requests and instead looks for input on stdin only from the frontend that initiated the debugging. Sending input other than keystrokes to the kernel when it is in debug will require generating kernel messages at a much lower level, and I'm not sure diving in to modify the Qt console inner workings is wise right now. As a simpler approach today, I would propose we live with the limitation that today the debug toolbar does not work for IPython shells. I have modified the debug action code to basically ignore the debug command actions for IPython shells.

I also relented and included a "Debug file" button in the debug toolbar. Inserting the button is a little inelegant since the run menu and associated actions are managed by the editor plugin. Hopefully it suffices for today.

I have pushed up the associated changes to my clone at bookmark jedludlow-issue-554 here: https://code.google.com/r/jedludlow-spyderlib-default/source/list?name=jedludlow-issue-554 Since this branch was initiated quite some time ago you will see several merges back *from* the main development line *onto* this branch to keep it up to date. The most recent was this morning. As a result, it will merge back onto the main development line trivially. I would request a final review and pull if acceptable. I've added Carlos and Pierre as CCs on the issue for this purpose.

**Cc:** ccordoba12 pierre.raybaut

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2012-12-26T21:51:15Z_

Jed, could you remove (or reduce) the merges with the main development line before merging the new toolbar? I think you could export as patches the commits where you added new code and create a new branch just with them.

The thing is, as it is now, your work adds too much noise to the mercurial history. This is what I see in tortoisehg, because for every merge it creates a new branch in the log graph.

Another problem is your new use of F11, which is giving a message about a keyboard shortcut conflict when one opens the Preferences dialog. Did you see it? What a user should do to avoid it?

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2012-12-27T06:48:52Z_

Carlos, at your suggestion I have made the following changes and propose them for your review:

1. I rebased all of the relevant debugging toolbar commits off of the latest tip of the default branch. These should be easier to review now. I had originally chosen to keep the debugging toolbar branch up to date using periodic merging from the default tip (see [1] below) since I had published it publicly so long ago. But the rebase was fairly simple, so no problem there.

2. I changed all the debugging hot key additions to key combinations that were previously unused, so no conflicts with previous defaults should arise. I had asked the mailing list for feedback on re-mapping some of the existing hot keys (see [2]), but I didn't get much response. I guess I figured that meant no one was opposed to the remapping :). No worries. We'll just go with a default set of mappings that does not conflict with any previous default mappings, and each user can ultimately set these to whatever they prefer anyway. As a general note, we run the risk of generating a hot key conflict any time we introduce a new hot key even if the new binding does not conflict with previous defaults since any given user could already have a customized binding that will conflict with the new one we are introducing. This seems unavoidable today unless we introduce a more complex system that prioritizes user customizations over the defaults. Fortunately, the preferences dialog handles conflicts in a reasonably elegant way today.

The new rebased changes can be pulled from the bookmark jedludlow-issue-554-rebased in my clone here: https://code.google.com/r/jedludlow-spyderlib-default/source/list?name=jedludlow-issue-554-rebased I have closed my earlier debugging toolbar branch related to this issue so that no future changes will be inadvertently pulled from there.

\----------------------------------------

[1] http://stackoverflow.com/questions/2672351/hg-how-to-do-a-rebase-like-gits-rebase/2672489#2672489 [2] https://groups.google.com/forum/#!msg/spyderlib/udt8kfUyxJ0/KzN80rBOgKAJ
